File: Benchmarks/Core/Scene/EntityManager.cs
using Benchmarks.Core.Collections;
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;

namespace Benchmarks.Core.Scene
{
    /// <summary>
    /// Clase que administra las entidades de una escena.
    /// Replica exacta del patron de parallel arrays del engine.
    /// </summary>
    public class EntityManager(YTB<Yotsuba> entities)
    {
        public YTB<Yotsuba> YotsubaEntities = entities;

        public YTB<TransformComponent> TransformComponents = new YTB<TransformComponent>();
        public YTB<SpriteComponent2D> Sprite2DComponents = new YTB<SpriteComponent2D>();
        public YTB<AnimationComponent2D> Animation2DComponents = new YTB<AnimationComponent2D>();
        public YTB<RigidBodyComponent2D> Rigidbody2DComponents = new YTB<RigidBodyComponent2D>();
        public YTB<InputComponent> InputComponents = new YTB<InputComponent>();
        public YTB<ModelComponent3D> ModelComponents3D = new YTB<ModelComponent3D>();
        public YTB<TileMapComponent2D> TileMapComponent2Ds = new YTB<TileMapComponent2D>();
        public YTB<ButtonComponent2D> Button2DComponents = new YTB<ButtonComponent2D>();
        public YTB<ScriptComponent> ScriptComponents = new YTB<ScriptComponent>();
        public YTB<FontComponent2D> Font2DComponents = new YTB<FontComponent2D>();
        public YTB<ShaderComponent> ShaderComponents = new YTB<ShaderComponent>();
        public YTB<YTBModelComponent3D> YtbModelComponents = new YTB<YTBModelComponent3D>();

        public int CameraEntityId { get; set; } = -1;
        public float CameraZoom { get; set; } = 1.0f;

        public EntityManager() : this(new YTB<Yotsuba>())
        {
        }

        public void AddEntity(ref Yotsuba entity)
        {
            int Index = YotsubaEntities.Add(entity);
            YotsubaEntities[Index].Id = Index;
            entity.Id = Index;
            int ComponentIndex = 0;

            if (TransformComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(TransformComponent)}. Index {Index} != {ComponentIndex}");

            if (Sprite2DComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(SpriteComponent2D)}. Index {Index} != {ComponentIndex}");

            if (Animation2DComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(AnimationComponent2D)}. Index {Index} != {ComponentIndex}");

            if (Rigidbody2DComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(RigidBodyComponent2D)}. Index {Index} != {ComponentIndex}");

            if (InputComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(InputComponent)}. Index {Index} != {ComponentIndex}");

            if (ModelComponents3D.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(ModelComponent3D)}. Index {Index} != {ComponentIndex}");

            if (Button2DComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(ButtonComponent2D)}. Index {Index} != {ComponentIndex}");

            if (ScriptComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(ScriptComponent)}. Index {Index} != {ComponentIndex}");

            if (TileMapComponent2Ds.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(TileMapComponent2D)}. Index {Index} != {ComponentIndex}");

            if (Font2DComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(FontComponent2D)}. Index {Index} != {ComponentIndex}");

            if (ShaderComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(ShaderComponent)}. Index {Index} != {ComponentIndex}");

            if (YtbModelComponents.Add(default, out ComponentIndex) != Index)
                throw new InvalidOperationException(
                    $"Component entered at different index than entity. Component: {typeof(YTBModelComponent3D)}. Index {Index} != {ComponentIndex}");
        }

        public void AddTransformComponent(Yotsuba entity, TransformComponent component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Transform);
            TransformComponents[(uint)entity.Id] = component;
        }

        public void AddSpriteComponent(Yotsuba entity, SpriteComponent2D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Sprite);
            Sprite2DComponents[(uint)entity.Id] = component;
        }

        public void AddAnimationComponent(Yotsuba entity, AnimationComponent2D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Animation);
            Animation2DComponents[(uint)entity.Id] = component;
        }

        public void AddRigidbodyComponent(Yotsuba entity, RigidBodyComponent2D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Rigibody);
            Rigidbody2DComponents[(uint)entity.Id] = component;
        }

        public void AddInputComponent(Yotsuba entity, InputComponent component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Input);
            InputComponents[(uint)entity.Id] = component;
        }

        public void AddButtonComponent2D(Yotsuba entity, ButtonComponent2D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Button2D);
            Button2DComponents[(uint)entity.Id] = component;
        }

        public void AddShaderComponent2D(Yotsuba entity, ShaderComponent component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Shader);
            ShaderComponents[(uint)entity.Id] = component;
        }

        public void AddScriptComponent(Yotsuba entity, ScriptComponent script)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Script);
            ScriptComponents[(uint)entity.Id] = script;
        }

        public void AddFontComponent2D(Yotsuba entity, FontComponent2D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Font);
            Font2DComponents[(uint)entity.Id] = component;
        }

        public void AddTileMapComponent(Yotsuba entity, TileMapComponent2D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.TileMap);
            TileMapComponent2Ds[(uint)entity.Id] = component;
        }

        public void AddModelComponent3D(Yotsuba entity, ModelComponent3D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.Model3D);
            ModelComponents3D[(uint)entity.Id] = component;
        }

        public void AddYTBObject3D(Yotsuba entity, YTBModelComponent3D component)
        {
            YotsubaEntities[entity.Id].AddComponent(YTBComponent.YTBModel3D);
            YtbModelComponents[(uint)entity.Id] = component;
        }
    }
}

File: Benchmarks_FSharp/Core/Scene/EntityManager.fs
namespace Benchmarks.Core.Scene

open System
open Benchmarks.Core.Collections
open Benchmarks.Core.Components
open Benchmarks.Core.Entity

type EntityManager(entities: YTB<Yotsuba>) =
    member val YotsubaEntities = entities with get, set

    member val TransformComponents = YTB<TransformComponent>() with get, set
    member val Sprite2DComponents = YTB<SpriteComponent2D>() with get, set
    member val Animation2DComponents = YTB<AnimationComponent2D>() with get, set
    member val Rigidbody2DComponents = YTB<RigidBodyComponent2D>() with get, set
    member val InputComponents = YTB<InputComponent>() with get, set
    member val ModelComponents3D = YTB<ModelComponent3D>() with get, set
    member val TileMapComponent2Ds = YTB<TileMapComponent2D>() with get, set
    member val Button2DComponents = YTB<ButtonComponent2D>() with get, set
    member val ScriptComponents = YTB<ScriptComponent>() with get, set
    member val Font2DComponents = YTB<FontComponent2D>() with get, set
    member val ShaderComponents = YTB<ShaderComponent>() with get, set
    member val YtbModelComponents = YTB<YTBModelComponent3D>() with get, set

    member val CameraEntityId = -1 with get, set
    member val CameraZoom = 1.0f with get, set

    new() = EntityManager(YTB<Yotsuba>())

    member this.AddEntity(entity: byref<Yotsuba>) =
        let index = this.YotsubaEntities.Add(entity)
        this.YotsubaEntities.[index].Id <- index
        entity.Id <- index
        let mutable componentIndex = 0

        if this.TransformComponents.Add(Unchecked.defaultof<TransformComponent>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<TransformComponent>}. Index {index} != {componentIndex}"))
        if this.Sprite2DComponents.Add(Unchecked.defaultof<SpriteComponent2D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<SpriteComponent2D>}. Index {index} != {componentIndex}"))
        if this.Animation2DComponents.Add(Unchecked.defaultof<AnimationComponent2D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<AnimationComponent2D>}. Index {index} != {componentIndex}"))
        if this.Rigidbody2DComponents.Add(Unchecked.defaultof<RigidBodyComponent2D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<RigidBodyComponent2D>}. Index {index} != {componentIndex}"))
        if this.InputComponents.Add(Unchecked.defaultof<InputComponent>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<InputComponent>}. Index {index} != {componentIndex}"))
        if this.ModelComponents3D.Add(Unchecked.defaultof<ModelComponent3D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<ModelComponent3D>}. Index {index} != {componentIndex}"))
        if this.Button2DComponents.Add(Unchecked.defaultof<ButtonComponent2D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<ButtonComponent2D>}. Index {index} != {componentIndex}"))
        if this.ScriptComponents.Add(Unchecked.defaultof<ScriptComponent>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<ScriptComponent>}. Index {index} != {componentIndex}"))
        if this.TileMapComponent2Ds.Add(Unchecked.defaultof<TileMapComponent2D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<TileMapComponent2D>}. Index {index} != {componentIndex}"))
        if this.Font2DComponents.Add(Unchecked.defaultof<FontComponent2D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<FontComponent2D>}. Index {index} != {componentIndex}"))
        if this.ShaderComponents.Add(Unchecked.defaultof<ShaderComponent>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<ShaderComponent>}. Index {index} != {componentIndex}"))
        if this.YtbModelComponents.Add(Unchecked.defaultof<YTBModelComponent3D>, &componentIndex) <> index then
            raise (InvalidOperationException($"Component entered at different index than entity. Component: {typeof<YTBModelComponent3D>}. Index {index} != {componentIndex}"))

    member this.AddTransformComponent(entity: Yotsuba, component': TransformComponent) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Transform)
        this.TransformComponents.SetByUint(uint entity.Id, component')

    member this.AddSpriteComponent(entity: Yotsuba, component': SpriteComponent2D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Sprite)
        this.Sprite2DComponents.SetByUint(uint entity.Id, component')

    member this.AddAnimationComponent(entity: Yotsuba, component': AnimationComponent2D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Animation)
        this.Animation2DComponents.SetByUint(uint entity.Id, component')

    member this.AddRigidbodyComponent(entity: Yotsuba, component': RigidBodyComponent2D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Rigibody)
        this.Rigidbody2DComponents.SetByUint(uint entity.Id, component')

    member this.AddInputComponent(entity: Yotsuba, component': InputComponent) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Input)
        this.InputComponents.SetByUint(uint entity.Id, component')

    member this.AddButtonComponent2D(entity: Yotsuba, component': ButtonComponent2D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Button2D)
        this.Button2DComponents.SetByUint(uint entity.Id, component')

    member this.AddShaderComponent2D(entity: Yotsuba, component': ShaderComponent) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Shader)
        this.ShaderComponents.SetByUint(uint entity.Id, component')

    member this.AddScriptComponent(entity: Yotsuba, script: ScriptComponent) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Script)
        this.ScriptComponents.SetByUint(uint entity.Id, script)

    member this.AddFontComponent2D(entity: Yotsuba, component': FontComponent2D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Font)
        this.Font2DComponents.SetByUint(uint entity.Id, component')

    member this.AddTileMapComponent(entity: Yotsuba, component': TileMapComponent2D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.TileMap)
        this.TileMapComponent2Ds.SetByUint(uint entity.Id, component')

    member this.AddModelComponent3D(entity: Yotsuba, component': ModelComponent3D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.Model3D)
        this.ModelComponents3D.SetByUint(uint entity.Id, component')

    member this.AddYTBObject3D(entity: Yotsuba, component': YTBModelComponent3D) =
        this.YotsubaEntities.[entity.Id].AddComponent(YTBComponent.YTBModel3D)
        this.YtbModelComponents.SetByUint(uint entity.Id, component')
